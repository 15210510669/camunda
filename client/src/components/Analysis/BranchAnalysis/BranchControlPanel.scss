.BranchControlPanel {
  background-color: var(--silver-base-97);
  border-radius: 8px;
  border: 1px solid var(--silver-darken-87);
  padding: 12px;
  margin-bottom: 1em;

  .list {
    list-style: none;
    margin: 0;
    padding: 0;
    display: inline;

    .item {
      display: inline;
      position: relative;
      padding-right: 20px;

      &.summary {
        line-height: 50px;

        .DefinitionSelection {
          display: inline-block;
          margin: 0 0.5em;
          line-height: normal;
          width: 220px;

          .Popover .buttonWrapper .Button {
            padding-left: 20px;
          }
        }
      }

      .config {
        cursor: default;
        display: inline-block;
        margin: 0 0.5em;
        line-height: 20px;
      }
    }

    .Filter {
      .filterHeader {
        position: relative;
        display: inline-block;
        margin-right: 20px;

        .dropdownLabel {
          margin-right: 10px;
          flex-grow: 1;
        }

        .explanation {
          display: none;
        }
      }

      .emptyMessage,
      .linkingTip {
        display: none;
      }

      .FilterList {
        display: inline;

        > .listItem {
          display: inline-flex;
          width: auto;
          margin-right: 10px;
        }
      }
    }
  }
}
