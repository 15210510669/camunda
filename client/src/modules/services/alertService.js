/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import {get, post, put, del} from 'request';

export async function loadAlerts(collection) {
  const response = await get(`api/collection/${collection}/alerts`);
  return await response.json();
}

export async function addAlert(alert) {
  return await post(`api/alert`, alert);
}

export async function editAlert(id, alert) {
  return await put(`api/alert/${id}`, alert);
}

export async function removeAlert(id) {
  return await del(`api/alert/${id}`);
}
