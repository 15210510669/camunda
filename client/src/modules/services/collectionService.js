/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import {removeEntityFromCollection, addEntityToCollection} from './entityService';

export function toggleEntityCollection(loadCollections) {
  return async function(entity, collection, isRemove) {
    if (isRemove) {
      await removeEntityFromCollection(entity.id, collection.id);
    } else {
      await addEntityToCollection(entity.id, collection.id);
    }

    await loadCollections();
  };
}

export function getEntitiesCollections(collections) {
  const entitiesCollections = {};

  collections.forEach(collection => {
    collection.data.entities.forEach(entity => {
      if (entitiesCollections[entity.id]) {
        entitiesCollections[entity.id].push(collection);
      } else {
        entitiesCollections[entity.id] = [collection];
      }
    });
  });

  return entitiesCollections;
}
