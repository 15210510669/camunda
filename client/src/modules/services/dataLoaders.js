/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import {get, post} from 'request';

export async function getFlowNodeNames(processDefinitionKey, processDefinitionVersion, tenantId) {
  if (processDefinitionKey && processDefinitionVersion) {
    const payload = {
      processDefinitionKey,
      processDefinitionVersion
    };

    if (tenantId) {
      payload.tenantId = tenantId;
    }

    const response = await post(`api/flow-node/flowNodeNames`, payload);

    const json = await response.json();

    return await json.flowNodeNames;
  } else {
    return {};
  }
}

export async function loadDefinitions(type, collectionId) {
  const params = {};
  if (collectionId) {
    params.filterByCollectionScope = collectionId;
  }

  const response = await get(`api/definition/${type}/definitionVersionsWithTenants`, params);

  return await response.json();
}

export async function loadProcessDefinitionXml(key, version, tenantId) {
  const payload = {key, version};
  if (tenantId) {
    payload.tenantId = tenantId;
  }

  try {
    const response = await get('api/definition/process/xml', payload);

    return await response.text();
  } catch (e) {
    return null;
  }
}

export async function loadDecisionDefinitionXml(key, version, tenantId) {
  const payload = {key, version};
  if (tenantId) {
    payload.tenantId = tenantId;
  }
  try {
    const response = await get('api/definition/decision/xml', payload);

    return await response.text();
  } catch (e) {
    return null;
  }
}

export async function checkDeleteConflict(id, entity) {
  const response = await get(`api/${entity}/${id}/delete-conflicts`);
  return await response.json();
}

const loadVariablesFrom = endpoint => async payload => {
  const response = await post(endpoint, payload);

  return await response.json();
};

export const loadVariables = loadVariablesFrom('api/variables');
export const loadInputVariables = loadVariablesFrom('api/decision-variables/inputs/names');
export const loadOutputVariables = loadVariablesFrom('api/decision-variables/outputs/names');
