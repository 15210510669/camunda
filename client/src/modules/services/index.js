import * as formattersImport from './formatters';
import * as reportConfigImport from './reportConfig';

export {getFlowNodeNames} from './GetFlowNodeNames';
export {extractProcessDefinitionName} from './GetProcessDefinitionName';
export {loadProcessDefinitions} from './LoadProcessDefinitions';
export {default as processRawData} from './processRawData';
export {numberParser} from './NumberParser';
export {isDurationValue} from './isDurationValue';
export {incompatibleFilters} from './incompatibleFilters';
export {loadEntity, checkDeleteConflict} from './entityServices';

export {getDiagramElementsBetween} from './diagramServices';
export {getCustomReportInfo, getCustomDashboardInfo} from './customEntityListServices';

// unfortunately, there is no syntax like "export * as formatters from './formatters'"
export const formatters = formattersImport;
export const reportConfig = reportConfigImport;
