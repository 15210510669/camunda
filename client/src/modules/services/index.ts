/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

export function getRandomId() {
  return Math.random().toString(36).slice(2);
}

export {TEXT_REPORT_MAX_CHARACTERS, isTextReportTooLong, isTextReportValid} from './reportService';
export {default as getScreenBounds} from './getScreenBounds';
export {default as ignoreFragments} from './ignoreFragments';
export {default as isReactElement} from './isReactElement';
