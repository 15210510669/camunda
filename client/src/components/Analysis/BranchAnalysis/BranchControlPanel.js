/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import update from 'immutability-helper';
import React from 'react';
import equal from 'deep-equal';
import {ActionItem, DefinitionSelection} from 'components';

import {getFlowNodeNames} from 'services';

import {Filter} from '../../Reports';

import './BranchControlPanel.scss';
import {t} from 'translation';

export default class BranchControlPanel extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      flowNodeNames: null
    };
  }

  componentDidMount() {
    this.loadFlowNodeNames();
  }

  loadFlowNodeNames = async () => {
    this.setState({
      flowNodeNames: await getFlowNodeNames(
        this.props.processDefinitionKey,
        this.props.processDefinitionVersions[0],
        this.props.tenantIds[0]
      )
    });
  };

  componentDidUpdate({processDefinitionKey, processDefinitionVersions}) {
    if (
      this.props.processDefinitionKey !== processDefinitionKey ||
      !equal(this.props.processDefinitionVersions, processDefinitionVersions)
    ) {
      this.loadFlowNodeNames();
    }
  }

  getDefinitionConfig = () => {
    return {
      processDefinitionKey: this.props.processDefinitionKey,
      processDefinitionVersions: this.props.processDefinitionVersions,
      tenantIds: this.props.tenantIds
    };
  };

  hover = element => () => {
    this.props.updateHover(element);
  };

  isProcDefSelected = () => {
    return this.props.processDefinitionKey && this.props.processDefinitionVersions;
  };

  renderInput = ({type, bpmnKey}) => {
    const {hoveredControl, hoveredNode} = this.props;
    const disableFlowNodeSelection = !this.isProcDefSelected();

    return (
      <div
        className="config"
        name={type}
        onMouseOver={this.hover(type)}
        onMouseOut={this.hover(null)}
      >
        <ActionItem
          disabled={disableFlowNodeSelection}
          onClick={() => this.props.updateSelection(type, null)}
          highlighted={
            !disableFlowNodeSelection &&
            (hoveredControl === type || (hoveredNode && hoveredNode.$instanceOf(bpmnKey)))
          }
        >
          {this.props[type]
            ? this.props[type].name || this.props[type].id
            : t(`analysis.emptySelectionLabel.${type}`)}
        </ActionItem>
      </div>
    );
  };

  render() {
    return (
      <div className="BranchControlPanel">
        <ul className="list">
          <li className="item summary">
            {t('analysis.selectLabel')}
            <DefinitionSelection
              type="process"
              definitionKey={this.props.processDefinitionKey}
              versions={this.props.processDefinitionVersions}
              tenants={this.props.tenantIds}
              xml={this.props.xml}
              onChange={({key, versions, tenantIds}) =>
                this.props.onChange({
                  processDefinitionKey: key,
                  processDefinitionVersions: versions,
                  tenantIds,
                  filter: this.props.filter.filter(
                    ({type}) =>
                      !['executedFlowNodes', 'executingFlowNodes', 'variable'].includes(type)
                  )
                })
              }
            />
            {t('analysis.gatewayLabel')}
            {this.renderInput({type: 'gateway', bpmnKey: 'bpmn:Gateway'})}
            {t('analysis.endEventLabel')}
            {this.renderInput({type: 'endEvent', bpmnKey: 'bpmn:EndEvent'})}
          </li>
          <li className="item itemFilter">
            <Filter
              data={this.props.filter}
              flowNodeNames={this.state.flowNodeNames}
              onChange={({filter}) =>
                this.props.onChange({filter: update(this.props.filter, filter)})
              }
              xml={this.props.xml}
              {...this.getDefinitionConfig()}
            />
          </li>
        </ul>
      </div>
    );
  }
}
