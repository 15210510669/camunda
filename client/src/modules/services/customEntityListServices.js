import {extractProcessDefinitionName} from 'services';

export function getCustomReportInfo(report) {
  if (report.data) {
    // if not empty combined
    if (report.reportType === 'combined' && report.data.reportIds && report.data.reportIds.length) {
      const reportsCount = report.data.reportIds.length;
      return `${reportsCount} report${reportsCount !== 1 ? 's' : ''}`;
    }
    // if normal report
    if (report.data.configuration.xml) {
      return extractProcessDefinitionName(report.data.configuration.xml);
    }
  }
  return '';
}

export function getCustomDashboardInfo(dashboard) {
  if (dashboard.reports.length === 1) return '1 Report';
  return dashboard.reports.length + ' Reports';
}
