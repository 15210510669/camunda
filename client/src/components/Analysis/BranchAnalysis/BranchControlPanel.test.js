/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import React from 'react';
import {shallow} from 'enzyme';

import {loadVariables} from 'services';

import {BranchControlPanel} from './BranchControlPanel';

jest.mock('services', () => {
  return {
    ...jest.requireActual('services'),
    loadVariables: jest.fn().mockReturnValue([{name: 'variable1', type: 'String'}]),
  };
});

const data = {
  processDefinitionKey: 'aKey',
  processDefinitionVersions: ['aVersion'],
  tenantIds: [],
  filter: null,
  xml: 'aFooXml',
  mightFail: jest.fn().mockImplementation((data, cb) => cb(data)),
};

const emptyData = {
  processDefinitionKey: '',
  processDefinitionVersions: [],
  tenantIds: [],
  filter: null,
  xml: null,
};

const spy = jest.fn();

it('should contain a gateway and end Event field', () => {
  const node = shallow(<BranchControlPanel {...data} onChange={spy} />);

  expect(node.find('[name="gateway"]')).toExist();
  expect(node.find('[name="endEvent"]')).toExist();
});

it('should show a please select message if an entity is not selected', () => {
  const node = shallow(<BranchControlPanel {...data} onChange={spy} />);

  expect(node.find('SelectionPreview').at(0).dive()).toIncludeText('Select Gateway');
  expect(node.find('SelectionPreview').at(1).dive()).toIncludeText('Select End Event');
});

it('should show the element name if an element is selected', () => {
  const node = shallow(
    <BranchControlPanel
      {...data}
      onChange={spy}
      gateway={{
        name: 'I am a Gateway',
        id: 'gatewayId',
      }}
    />
  );

  expect(node.find('SelectionPreview').at(0).dive()).toIncludeText('I am a Gateway');
  expect(node.find('SelectionPreview').at(0).dive()).not.toIncludeText('gatewayId');
});

it('should show the element id if an element has no name', () => {
  const node = shallow(
    <BranchControlPanel
      {...data}
      onChange={spy}
      gateway={{
        name: undefined,
        id: 'gatewayId',
      }}
    />
  );

  expect(node.find('SelectionPreview').at(0).dive()).toIncludeText('gatewayId');
});

it('should disable gateway and EndEvent elements if no ProcDef selected', async () => {
  const node = await shallow(<BranchControlPanel hoveredControl="gateway" {...emptyData} />);

  expect(node.find('SelectionPreview').at(0)).toBeDisabled();
  expect(node.find('SelectionPreview').at(1)).toBeDisabled();

  expect(node.find('SelectionPreview').at(1)).not.toHaveClassName(
    'BranchControlPanel__config--hover'
  );
});

it('should pass the xml to the Filter component', async () => {
  const node = await shallow(<BranchControlPanel {...data} />);
  const filter = node.find('Filter');
  expect(filter.find('[xml="aFooXml"]')).toExist();
});

it('should load the variable names and hand them to the filter if process definition changes', async () => {
  const node = shallow(<BranchControlPanel {...data} />);
  node.setProps({
    processDefinitionKey: 'fooKey',
    processDefinitionVersions: ['fooVersion'],
  });

  await flushPromises();
  node.update();

  expect(loadVariables).toHaveBeenCalled();
  expect(node.find('Filter').prop('variables')).toEqual(loadVariables());
});

it('should display a sentence to describe what the user can do on this page', () => {
  const node = shallow(<BranchControlPanel {...emptyData} />);

  expect(node).toMatchSnapshot();
});
