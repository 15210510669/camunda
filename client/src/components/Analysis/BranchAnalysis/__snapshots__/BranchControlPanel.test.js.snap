// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display a sentence to describe what the user can do on this page 1`] = `
<div
  className="BranchControlPanel"
>
  <ul
    className="list"
  >
    <li
      className="item summary"
    >
      For
      <withRouter(DefinitionSelectionErrorHandler)
        definitionKey=""
        onChange={[Function]}
        tenants={Array []}
        type="process"
        versions={Array []}
        xml={null}
      />
      analyze how the branches of
      <div
        className="config"
        name="gateway"
        onMouseOut={[Function]}
        onMouseOver={[Function]}
      >
        <ActionItem
          disabled={true}
          highlighted={false}
          onClick={[Function]}
        >
          Select Gateway
        </ActionItem>
      </div>
      affect the probability that an instance reached
      <div
        className="config"
        name="endEvent"
        onMouseOut={[Function]}
        onMouseOver={[Function]}
      >
        <ActionItem
          disabled={true}
          highlighted={false}
          onClick={[Function]}
        >
          Select End Event
        </ActionItem>
      </div>
    </li>
    <li
      className="item itemFilter"
    >
      <Filter
        data={null}
        filterLevel="instance"
        flowNodeNames={null}
        onChange={[Function]}
        processDefinitionKey=""
        processDefinitionVersions={Array []}
        tenantIds={Array []}
        variables={null}
        xml={null}
      />
    </li>
  </ul>
</div>
`;
